Reword and merge 2 questions about Rust desiderata (#284)
rust-lang/surveys: surveys/2024-annual-survey/questions.md

### Which of the following aspects of Rust present non-trivial problems to your programming productivity?

> #### Choose an assessment only for challenges that you think are impacting your work. You can skip any topic, we will interpret that as you are neutral.

Author: ```suggestion
#### Assess only challenges that you think are impacting your work.
```

Type: matrix (optional)

- Implementing logic for tuples of various sizes
- Splitting code across crates (orphan rule)
- Having to implement Iterator manually
- Not being able to do enough in const fn
- Implementing plugins
- Interoperating with other languages (e.g. C or C++)
- Achieving structured concurrency with async code
- Writing executor-agnostic async code
- Writing correct unsafe code
- Borrow checker not allowing valid code
- Slow runtime performance
- Slow compile times
- Large binary size of compiled artifacts
- High disk space usage (e.g. the size of the target folder)
- Encountering compiler bugs (e.g. ICEs a.k.a. internal compiler errors or miscompilations)
- Encountering opaque/unclear compiler error messages
- Subpar IDE support (e.g. some errors are not shown or analysis is slow)
- Subpar debugging experience (e.g. missing value visualizations or async stacktraces)
- Lacking documentation of the Rust language or the standard library

Options:

- Big problem for me

> - Could be improved, but does not limit me

Reviewer: It's severe bikeshedding at this point :laughing: But I think that this changes the meaning quite a lot.

The important message that I want to get out of this is whether some challenge is a limitation/blocker/big problem or rather "just an annoyance". This second answer should have been "just an annoyance".

Saying "it does not affect me" overlaps with the third option, and thus there's no option that would state "I see this as a problem, but it's not *that* bad". For example, there are things that probably affect almost everyone (like compile times), that everyone wants to get improved (who woud want *longer* compile times?), but for some people it's not really a blocker for their productivity.

> - Could be improved, but does not limit me

Author: agreed. thanks for clarifying the point (which I had missed)

- Not an issue for me at all

> **justification**
>
> This question tries to get insights on what people wish the Rust project would improve.

> ### Are there any challenges not mentioned above that limit your Rust programming productivity?

Author: ```suggestion
### Are there any challenges not mentioned above that affect your Rust programming productivity?
```

- Achieving structured concurrency with async code
- Writing executor-agnostic async code
- Borrow checker not allowing valid code
- Slow runtime performance
- Slow compile times
- Achieving small binary size of compiled artifacts
- Memory usage (example: how much RAM rustc uses when compiling)
- Disk space usage (example: the size of target folder)
- Bugs in the compiler (that is, ICEs a.k.a. internal compiler errors, miscompilations, etc.)
- Opaque / unclear compiler error messages
- Subpar IDE support
- Debugging experience is not good enough (example: unclear / misleading diagnostics)
- Package management (crates.io)
- Rust language and standard library documentation

> How bad do I need this?:

Reviewer: There isn't actually any label for this in SurveyHero, so this text ("How bad do I need this") won't appear anywhere. Which is good, because the wording would be weird ("Slow compile times" -> "How bad do I need this" is probably not a good lead-in :laughing: ). We could add some text just below the question text though.

So either we have to ask the main question in a way that the prepared answers make sense for it, or explain the answers below the question.

It seems to me that the answers in the vein of "it would do something for me" are too.. complicated, so that they make sense across all the options. What about something like this? It's hard to come up with good options for all four columns though..
![image](https://github.com/user-attachments/assets/96a14d67-7b88-4617-acea-f71a71f3ada4)

Author: How does SH recognize and separate in the matrix the block of answer options from the topics? "How bad do I need this" was just a funny block separator between the two but it can be removed (and def. should be reworded).

Choose an assessment only for the topics you think they are impacting your work. If you don't choose an answer, we will interpret as "I don't know" or that you are neutral on that topic.

About the column layout: now that I see it in the layout I'd say let's remove the "I don't know" option to win some horizontal space an try to reword the answer options as you suggest to keep them as short as possible.

Reviewer: > How does SH recognize and separate in the matrix the block of answer options from the topics? "How bad do I need this" was just a funny block separator between the two but it can be removed (and def. should be reworded).

It doesn't, we fill in the questions into SH manually xD We do tend to include *some* separator in the Markdown file though, to recognize the split, as you said.

> About the column layout: now that I see it in the layout I'd say let's remove the "I don't know" option to win some horizontal space an try to reword the answer options as you suggest to keep them as short as possible.

Ok, we can just interpret a missing answer as "I don't know" (as the text already says), it will just need a bit of postprocessing.

Author: I have pushed a proposal to drastically shorten the colum width? How does it look?

Reviewer: Having just three columns is :+1: But yeah, I'll need to think of the wording to unify it. "slow runtime performance" -> "very much needed" isn't ideal xD I'll try to rephrase all questions and the answers so that it makes more sense.

Author: > I'll try to rephrase all questions and the answers so that it makes more sense.

I'll also have a look at that

- Achieving structured concurrency with async code
- Writing executor-agnostic async code
- Borrow checker not allowing valid code

- (From here the additional one)

> - Runtime performance

Author: This is also a blanket statement I think it's not very useful. Does it make sense to say "faster!!11" without specifying where/when/what?

Reviewer: I think that we can reword everything to be a "challenge" or a "problem" (since we're looking for problems and roadblocks, rather than solutions). So having "slow runtime performance" seems fine to me.

> - Compile times

Reviewer: In the same vein, I'd write something like "Compilation is too slow" (we could even split it into incremental vs from scratch builds).

> - Binary size (of the final compiled artifact)

Reviewer: "Achieving small binary size of compiled artifacts" or "Rust binaries are too large".

> - Memory usage (example: how much RAM rustc uses when compiling)

Reviewer: Memory usage of the Rust compiler or the compiled Rust programs.

Author: but these are two different questions, right?

> - Disk space usage (example: the size of target folder)

Reviewer: `Disk space usage is too high"?

Author: :shrug:

- Bugs in the compiler (that is, ICEs a.k.a. internal compiler errors, miscompilations, etc.)
- Opaque / unclear compiler error messages

> - IDE experience

Author: IDE is a bit vague (I use emacs, you use vstudio). If this question means rust analyzer maybe explicit it?

Reviewer: I would reword to something like "Insufficient IDE support" or "Rust IDE support is not good enough or is too slow", or something like that.

Author: I would keep the wording neutral because the quality assessment (good/bad) is up to the surveyer.

I was just wondering if it makes sense to coalesce all IDEs together, each of them has a completely different story and lifecycle. Maybe I use X and it's a stellar experience and you use Y and it's bad, it's apple and oranges and these differences are lost here.

Well, let's leave it like that, there's a specific question about "which IDE do you use" :man_shrugging:

> - Debugging experience (example: unclear / misleading diagnostics)

Reviewer: Debugging experience is not good enough.

We could even add another one for async debugging, but that's probably too detailed (and we know anyway that async debugging in Rust sucks now, so.. :) ).

- Documentation (rustdoc, docs.rs)

> - Build system (cargo)

Reviewer: Building Rust projects?

Reviewer: Or we can just scrap this. I just wanted to avoid the situation from last year, where we asked "which Rust aspects are causing you issues" and one of the answers was just "async" or "macros" xD

> - Package management (crates.io)

Reviewer: Using dependencies? Or e.g. auditing dependencies? We can just arbitrarily choose some common complaints :)

Author: Heh, this is also a wide topic. I don't know how to change this, what do we really want to point at

> - New language features

Author: This is also vague ("I want new things" - yeah but what?). I think we can remove it.

Reviewer: Yeah, I would remove it, along with the blanket statements, like documentation or build system. It should be specific things, like "Code/feature that I'm using lacks documentation" or something like that.

- Rust language and standard library documentation

> - Other (please specify)

Reviewer: Matrix questions in SH sadly cannot have the "Other" open category :( So we will need additional question anyway, to ask for open answers. This would mean that we wouldn't reduce question count by one, which is a bit annoying, but nothing too bad I guess.

How bad do I need this?:

- Would unblock my use-case
- Would greatly improve my developer experience

> - Not a priority / good enough for me

Reviewer: We should also add an option I don't know (or I don't care), and clarify that not choosing an answer equals to saying "I don't know".
